mqtt/init.lua

```diff
@@ -81,6 +81,192 @@ function mqtt.run_sync(cl)
 	end
 end
 
+
+--- Validates a topic with wildcards.
+-- @tparam string t wildcard topic to validate
+-- @return topic, or false+error
+-- @usage
+-- local t = "invalid/#/subscribe/#/topic"
+-- local topic = assert(mqtt.validate_subscribe_topic(t))
+function mqtt.validate_subscribe_topic(t)
+	if type(t) ~= "string" then
+		return false, "bad subscribe-topic; expected topic to be a string, got: "..type(t)
+	end
+	if #t < 1 then
+		return false, "bad subscribe-topic; expected minimum topic length of 1"
+	end
+	do
+		local _, count = t:gsub("#", "")
+		if count > 1 then
+			return false, "bad subscribe-topic; wildcard '#' may only appear once, got: '"..t.."'"
+		end
+		if count == 1 then
+			if t ~= "#" and not t:find("/#$") then
+				return false, "bad subscribe-topic; wildcard '#' must be the last character, and " ..
+				              "be prefixed with '/' (unless the topic is '#'), got: '"..t.."'"
+			end
+		end
+	end
+	do
+		local t1 = "/"..t.."/"
+		local i = 1
+		while i do
+			i = t1:find("+", i)
+			if i then
+				if t1:sub(i-1, i+1) ~= "/+/" then
+					return false, "bad subscribe-topic; wildcard '+' must be enclosed between '/' " ..
+					              "(except at start/end), got: '"..t.."'"
+				end
+				i = i + 1
+			end
+		end
+	end
+	return t
+end
+
+--- Validates a topic without wildcards.
+-- @tparam string t topic to validate
+-- @return topic, or false+error
+-- @usage
+-- local t = "invalid/#/publish/+/topic"
+-- local topic = assert(mqtt.validate_publish_topic(t))
+function mqtt.validate_publish_topic(t)
+	if type(t) ~= "string" then
+		return false, "bad publish-topic; expected topic to be a string, got: "..type(t)
+	end
+	if #t < 1 then
+		return false, "bad publish-topic; expected minimum topic length of 1"
+	end
+	if t:find("+", nil, true) or t:find("#", nil, true) then
+		return false, "bad publish-topic; wildcards '#', and '+' are not allowed when publishing, got: '"..t.."'"
+	end
+	return t
+end
+
+do
+	local MATCH_ALL = "(.+)"     -- matches anything at least 1 character long
+	local MATCH_HASH = "(.-)"    -- match anything, can be empty
+	local MATCH_PLUS = "([^/]-)" -- match anything between '/', can be empty
+
+	--- Returns a Lua pattern from topic.
+	-- Takes a wildcarded-topic and returns a Lua pattern that can be used
+	-- to validate if a received topic matches the wildcard-topic
+	-- @tparam string t the wildcard topic
+	-- @return Lua-pattern (string) or throws error on invalid input
+	-- @usage
+	-- local patt = compile_topic_pattern("homes/+/+/#")
+	--
+	-- local incoming_topic = "homes/myhome/living/mainlights/brightness"
+	-- local homeid, roomid, varargs = incoming_topic:match(patt)
+	function mqtt.compile_topic_pattern(t)
+		t = assert(mqtt.validate_subscribe_topic(t))
+		if t == "#" then
+			t = MATCH_ALL
+		else
+			t = t:gsub("#", MATCH_HASH)
+			t = t:gsub("%+", MATCH_PLUS)
+		end
+		return "^"..t.."$"
+	end
```

Comment:
Updated to return the number of fields and presence of varargs;
```suggestion
	function mqtt.compile_topic_pattern(t)
		t = assert(mqtt.validate_subscribe_topic(t))
		if t == "#" then
			return MATCH_ALL, 0, true
		end
		
		local fields, vararg
		t, vararg = t:gsub("#", MATCH_HASH)
		t, fields = t:gsub("%+", MATCH_PLUS)
		return "^"..t.."$", fields, vararg ~= 0
	end
```